src/components/GraphVisualization/GraphComponent.tsx

```diff
@@ -0,0 +1,122 @@
+import { Graph } from "react-d3-graph";
+import { SimpleGraph } from "./SimpleGraph";
+import { Node } from "./Node";
+
+export interface GraphData {
+  nodes: {
+    id: string;
+    x?: number;
+    y?: number;
+  }[];
+  links: {
+    source: string;
+    target: string;
+  }[];
+}
+
+const myConfig = {
+  nodeHighlightBehavior: true,
+  node: {
+    color: "lightgray",
+    highlightColor: "orange",
+    size: 200,
+  },
+  link: {
+    highlightColor: "orange",
+  },
+};
+
+function convertGraphDataToSimpleGraph(data: GraphData): SimpleGraph | undefined {
+  const graph = new SimpleGraph();
+  data.nodes.forEach((node) => {
+    graph.addNode(new Node(node.id));
+    return null;
+  });
+  data.links.forEach((link) => {
+    const fromNode = graph.findNodeById(link.source);
+    const toNode = graph.findNodeById(link.target);
+    if (fromNode === undefined || toNode === undefined) {
+      return undefined;
+    }
+    graph.addEdge(fromNode, toNode);
+    return graph;
+  });
+  return graph;
+}
+
+function allocateNodes(data: GraphData) {
+  const { links } = data;
+
+  const sources: Set<string> = new Set();
+  const targets: Set<string> = new Set();
+
+  links.forEach((link) => {
+    sources.add(link.source);
+    targets.add(link.target);
+  });
+
+  let root = "";
+  sources.forEach((source) => {
+    if (!targets.has(source)) {
+      root = source;
+    }
+  });
+
+  const simpleGraphFromData = convertGraphDataToSimpleGraph(data);
+  const depth = simpleGraphFromData?.dfsCountDepth(root);
+
+  // Hardcoded, intended to get the container size
+  const parentWidth = 500;
+  const parentHeight = 200;
+
+  let parentCellWidth = parentWidth;
+
+  for (let i = 0; i <= depth!; i += 1) {
+    const layer = simpleGraphFromData?.adjacencyList.filter((node) => node.height === i);
+    if (layer!.length > 1) {
+      layer?.sort((a: Node, b: Node) => a!.parent!.x - b!.parent!.x);
+    }
+    const layerY = (parentHeight / depth! / 2) * (1 + 2 * i);
+
+    const siblingsCount = layer!.length;
+
+    for (let j = 0; j < siblingsCount!; j += 1) {
+      let parentBorderX = 0;
+      if (layer![j]!.parent !== undefined) {
+        parentBorderX = layer![j]!.parent!.x - parentCellWidth / 2;
+      }
+      if (siblingsCount === 1) {
+        layer![j]!.x = parentBorderX + 0.5 * parentCellWidth;
+      } else {
+        layer![j]!.x =
+          parentBorderX +
+          ((2 * layer![j]!.parent!.getChildIndex(layer![j]!) + 1) * parentCellWidth) /
+            layer![j]!.parent!.children.length /
+            2;
+      }
+      layer![j]!.y = layerY;
+    }
+
+    if (siblingsCount > 1) {
+      parentCellWidth /= siblingsCount;
+    }
+  }
```

Comment:
:fearful:  :laughing: 
Myślę, że lepiej wynieść to do funkcji. 
Nie dlatego, że to może być gdzieś reużyte, a tylko dlatego, że lepiej wtedy się czyta kod.


Pewnie się zgodzisz, że lepiej by było gdyby funkcja to wyglądała tak:

```ts
const simpleGraphFromData = convertGraphDataToSimpleGraph(data);
const depth = simpleGraphFromData?.dfsCountDepth(root);
calculateCoordinates(simpleGraphFromData, depth); // tutaj migurje cała pętla
assignCoordinates(node, simpleGraphFromData) // tutaj migruje przypisywanie wartości
```